system/jlib/jmisc.hpp

```diff
@@ -334,4 +334,103 @@ extern jlib_decl char **getSystemEnv();
 
 extern jlib_decl char *getHPCCEnvVal(const char *name, const char *defaultValue);
 
+// class TimeDivisionTracker is useful for working out what a thread spends its time doing. See udptrrr.cpp for an example
+// of its usage
+
+template<unsigned NUMSTATES, bool reportOther> class TimeDivisionTracker
+{
+protected:
+    unsigned __int64 totals[NUMSTATES] = {0};
+    unsigned counts[NUMSTATES] = {0};
+    const char *stateNames[NUMSTATES];
+    unsigned __int64 lastTick = get_cycles_now();
+    unsigned currentState = 0;
+    StringAttr name;
+    unsigned __int64 reportIntervalCycles = 0;
+    unsigned __int64 lastReport = 0;
+
+    unsigned enterState(unsigned newState)
+    {
+        unsigned prevState = currentState;
+        unsigned __int64 now = get_cycles_now();
+        if (reportIntervalCycles && now - lastReport >= reportIntervalCycles)
+            report(true);
+        if (newState != currentState)
+        {
+            totals[currentState] += now - lastTick;
```

Comment:
If report() has been called, this will cause totals to be set to a  very high number - because lastTick will be set to the latest value of get_cycles_now() - which will be larger than 'now'.
Probably simplest to pass now into the report function, and have a public helper which does not require it to be passed (or similar).

Reply:
note: This will self correct later, so maybe not worth worrying about - but if so, it is worth adding a comment to clarify.